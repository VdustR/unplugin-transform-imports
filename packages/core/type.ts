export type Enforce = "post" | "pre" | null;
export type Cwd = string;
export type Transform = string | ((importName: string) => string);
export type Module = {
  path: string;
  transform?:
    | string
    | ((importName: string, moduleName: string, constName: string) => string);
};
export type Include = string;
export type Exclude = string;
export type TransformImportsOptions = {
  enforce?: Enforce;
  cwd?: Cwd;
  modules?: Module[];
  includes?: Include[];
  excludes?: Exclude[];
};
