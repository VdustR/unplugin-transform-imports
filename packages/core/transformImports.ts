import parser from "@babel/parser";
import type { ParserOptions } from "@babel/parser";
import traverse from "@babel/traverse";
import { transformFromAstAsync, types as t } from "@babel/core";
import type { TransformOptions } from "@babel/core";
import type { Module } from "./type";
import type {
  Identifier,
  ImportDeclaration,
  ImportDefaultSpecifier,
  ImportNamespaceSpecifier,
  ImportSpecifier,
  StringLiteral,
} from "@babel/types";

function isIdentifier(node: Identifier | StringLiteral): node is Identifier {
  return node.type === "Identifier";
}

export type ImportsTransformOptions = {
  parseOptions?: ParserOptions | undefined;
  transformOptions?: TransformOptions | undefined;
  modules: Module[];
};

export function groupImportSpecifier(
  specifiers: (
    | ImportDefaultSpecifier
    | ImportNamespaceSpecifier
    | ImportSpecifier
  )[]
): {
  importDefaultSpecifiers: ImportDefaultSpecifier[];
  importNamespaceSpecifiers: ImportNamespaceSpecifier[];
  importSpecifiers: ImportSpecifier[];
} {
  const importDefaultSpecifiers: ImportDefaultSpecifier[] = [];
  const importNamespaceSpecifiers: ImportNamespaceSpecifier[] = [];
  const importSpecifiers: ImportSpecifier[] = [];
  specifiers.forEach((specifier) => {
    if (specifier.type === "ImportDefaultSpecifier") {
      importDefaultSpecifiers.push(specifier);
    } else if (specifier.type === "ImportNamespaceSpecifier") {
      importNamespaceSpecifiers.push(specifier);
    } else {
      importSpecifiers.push(specifier);
    }
  });
  return {
    importDefaultSpecifiers,
    importNamespaceSpecifiers,
    importSpecifiers,
  };
}

export default async function importsTransform(
  code: string,
  { parseOptions, transformOptions, modules }: ImportsTransformOptions = {
    modules: [],
  }
) {
  const ast = parser.parse(code, { sourceType: "module", ...parseOptions });
  traverse(ast, {
    ImportDeclaration(path) {
      const moduleName = path.node.source.value;
      if (!moduleName) return;
      const matchedModule = modules.find((m) => m.path === moduleName);
      if (!matchedModule) return;
      if (!path.node.specifiers) return;
      const {
        importDefaultSpecifiers,
        importNamespaceSpecifiers,
        importSpecifiers,
      } = groupImportSpecifier(path.node.specifiers);
      // import * as xxx from "xxx";
      if (importNamespaceSpecifiers.length) return;
      // import xxx from "xxx";
      if (importSpecifiers.length === 0) return;
      const importDeclarations: ImportDeclaration[] = [];
      importDefaultSpecifiers.forEach((specifier) => {
        importDeclarations.push(
          t.importDeclaration([specifier], t.stringLiteral(moduleName))
        );
      });
      importSpecifiers.forEach((specifier) => {
        const [importName, constName] = [
          isIdentifier(specifier.imported)
            ? specifier.imported.name
            : specifier.imported.value,
          specifier.local.name,
        ] as [string, string];
        const transform = matchedModule.transform;
        const newModuleName = !transform
          ? `${moduleName}/${importName}`
          : typeof transform === "string"
          ? transform
              .replaceAll("${importName}", importName)
              .replaceAll("${moduleName}", moduleName)
              .replaceAll("${constName}", constName || importName)
          : transform(importName, moduleName, constName);
        importDeclarations.push(
          t.importDeclaration(
            [t.importDefaultSpecifier(t.identifier(constName))],
            t.stringLiteral(newModuleName)
          )
        );
      });
      path.replaceWithMultiple(importDeclarations);
    },
  });
  return transformFromAstAsync(ast, code, {
    babelrc: false,
    ...transformOptions,
  });
}
