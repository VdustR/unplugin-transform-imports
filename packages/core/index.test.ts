import transformImports from ".";
import { expect, it } from "vitest";
import { build } from "vite";
import { resolve } from "path";
import type { OutputChunk, RollupOutput } from "rollup";

it("Test build", async () => {
  const rollupOutput = await build({
    root: resolve(__dirname, "test"),
    configFile: false,
    plugins: [
      transformImports.vite({
        modules: [
          { path: "lodash" },
          {
            path: "@mui/icons-material",
            transform: "@mui/icons-material/$1",
          },
          {
            path: "phosphor-react",
            transform: (importName) => `phosphor-react/src/icons/${importName}`,
          },
          {
            path: "mdi-material-ui",
          },
        ],
      }),
    ],
    build: {
      minify: false,
      polyfillModulePreload: false,
      rollupOptions: {
        external: [
          /^lodash(\/.+)?/,
          /^@mui\/icons-material(\/.+)?/,
          /^phosphor-react(\/.+)?/,
          /^mdi-material-ui(\/.+)?/,
          /^react(\/.+)?/,
        ],
      },
    },
  });
  expect(
    (
      (rollupOutput as RollupOutput).output.find(
        ({ type }) => type === "chunk"
      ) as OutputChunk
    ).code
  ).toMatchSnapshot();
});
