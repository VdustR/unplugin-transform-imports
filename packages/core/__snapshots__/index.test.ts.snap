// Vitest Snapshot v1

exports[`Normal transform. > No transform 1`] = `
"import MdiCheck from \\"mdi-material-ui/Check\\";
import MuiCheck from \\"@mui/icons-material/Check\\";
import CheckBox from \\"@mui/icons-material/CheckBox\\";
import PhCheck from \\"phosphor-react/src/icons/Check\\";
import merge from \\"lodash/merge\\";
import lodash from \\"lodash\\";
import { useCallback, useMemo } from \\"react\\";
console.log({
  MdiCheck,
  MuiCheck,
  CheckBox,
  PhCheck,
  merge,
  lodash,
  useCallback,
  useMemo
});
"
`;

exports[`Test build > Test build 1`] = `
"import MdiCheck from \\"mdi-material-ui/Check\\";
import MuiCheck from \\"@mui/icons-material/Check\\";
import CheckBox from \\"@mui/icons-material/CheckBox\\";
import PhCheck from \\"phosphor-react/src/icons/Check\\";
import merge from \\"lodash/merge\\";
import lodash from \\"lodash\\";
import { useCallback, useMemo } from \\"react\\";
console.log({
  MdiCheck,
  MuiCheck,
  CheckBox,
  PhCheck,
  merge,
  lodash,
  useCallback,
  useMemo
});
"
`;

exports[`Test build 1`] = `
"import MdiCheck from \\"mdi-material-ui/Check\\";
import MuiCheck from \\"@mui/icons-material/Check\\";
import CheckBox from \\"@mui/icons-material/CheckBox\\";
import PhCheck from \\"phosphor-react/src/icons/Check\\";
import merge from \\"lodash/merge\\";
import lodash from \\"lodash\\";
import { useCallback, useMemo } from \\"react\\";
console.log({
  MdiCheck,
  MuiCheck,
  CheckBox,
  PhCheck,
  merge,
  lodash,
  useCallback,
  useMemo
});
"
`;
